### PR TITLE
:art: On the road to 0.3.0: Adding shift evaluation & more

### PR DESCRIPTION
**Pull request by alafage**
- :sparkles: Add evaluation of the metrics on a shifted dataset using the new `--eval_shift` argument 
- :sparkles: Add a corrupted dataset wrapper
- :sparkles: Add new Packed layers
- :sparkles: Add the focal loss
- :shirt: Plot segmentation results in the SegmentationRoutine logger.
- :shirt: Rework reliability diagrams for Calibration Error plot utils.
- ➖ Make some dependencies optional
- :bug: Fix corruption transforms

### CODE REVIEW

File: auto_tutorials_source/tutorial_bayesian.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def optim_lenet(model: nn.Module):
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 #
 # Now that we have prepared all of this, we just have to gather everything in
-# the main function and to train the model using the Lightning Trainer.
+# the main function and to train the model using the our wrapper of Lightning Trainer.
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_bayesian.py:114]: "the our"

File: auto_tutorials_source/tutorial_mc_batch_norm.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 First, we have to load the following utilities from TorchUncertainty:
 
-- the Trainer from Lightning
+- the TUTrainer from the framework
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_mc_batch_norm.py:16]: our framework

File: torch_uncertainty/datamodules/classification/cifar10.py [hunk 3]
```diff
@@ -74,17 +80,17 @@ def __init__(
         self.val_split = val_split
         self.num_dataloaders = num_dataloaders
         self.eval_ood = eval_ood
+        self.eval_shift = eval_shift
 
-        if test_alt == "c":
-            self.dataset = CIFAR10C
-        elif test_alt == "h":
+        if test_alt == "h":
```
Comment (o-laurent) [inline torch_uncertainty/datamodules/classification/cifar10.py:85]: Probably should check that the value is h or none

File: torch_uncertainty/datasets/corrupted.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
+from copy import deepcopy
```
Comment (o-laurent) [inline torch_uncertainty/datasets/corrupted.py:1]: There are no tests for this right now, I'll make them in post0

File: torch_uncertainty/routines/classification.py [hunk 5]
```diff
@@ -243,6 +250,11 @@ def _init_metrics(self) -> None:
             if self.eval_ood:
                 self.test_ood_ens_metrics = ens_metrics.clone(prefix="ood/ens_")
 
+            if self.eval_shift:
+                self.test_ood_ens_metrics = ens_metrics.clone(
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:254]: Fix this

File: torch_uncertainty/routines/segmentation.py [hunk 6]
```diff
@@ -231,17 +265,59 @@ def on_test_epoch_end(self) -> None:
...
+                "Selective Classification/Risk-Coverage curve",
                 self.test_sbsmpl_seg_metrics["sc/AURC"].plot()[0],
             )
             self.logger.experiment.add_figure(
-                "Generalized Risk-Coverage curve",
+                "Selective Classification/Generalized Risk-Coverage curve",
                 self.test_sbsmpl_seg_metrics["sc/AUGRC"].plot()[0],
             )
+            # Plot segmentation results.
+            for i, (img, pred, tgt) in enumerate(self.sample_buffer):
```
Comment (o-laurent) [inline torch_uncertainty/routines/segmentation.py:280]: split into a function?

Comment (alafage) [reply torch_uncertainty/routines/segmentation.py:280]: What do you mean?

Comment (o-laurent) [reply torch_uncertainty/routines/segmentation.py:280]: just a separate function to keep a clear on_test_epoch_end

**Review by o-laurent — APPROVED** [review]
